File: Day3/Day3.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    def traverse(self,root,path):
        if not root:
            path.append("None")
            return 
        path.append(str(root.val))
        self.traverse(root.left,path)
        self.traverse(root.right,path)
        
        
    def build(self,data):
        if data[0]=='None':
            data.pop(0)
            return 
        else:
            root=TreeNode(data[0])
            data.pop(0)
            root.left=self.build(data)
            root.right=self.build(data)
        return root
    
    
    def serialize(self, root):
        if root is None:
            return []
        path=[]
        self.traverse(root,path)
        path=','.join(path)
        return path
    
    
    def deserialize(self, data):
        if not data:
            return 
        data=data.split(',')
        return self.build(data)

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

File: Day3/Day3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    void traverse(TreeNode* root,ostringstream &os){
    if(root==NULL){
        os<<"# ";
        return;
    }
    os<<root->val<<" ";
    traverse(root->left,os);
    traverse(root->right,os);
}
    
    TreeNode* build(istringstream &is){
    string tmp;
    is>>tmp;
    if(tmp=="#") return NULL;
    auto root=new TreeNode(stoi(tmp));
    root->left=build(is);
    root->right=build(is);
    
    return root;
}

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream os;
    traverse(root,os);
    return os.str();
        
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream is(data);
    return build(is);
        
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));